### ISSUE TITLE
[Impeller] Support rendering compressed texture formats / add capability to detect support.

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#148443]
I _think_ that we could fairly easily add support for compressed image formats to Impeller, at least on the Vulkan backend to star with.

See: https://docs.vulkan.org/spec/latest/appendices/compressedtex.html

These formats are unlike PNG/JPEG and don't include all information required to process them, and would likely require a developer to provide a size/ format specification as well. For example something like:

```dart
import 'dart:ui' as uil

void main() {
  Uint8List data = getBytes(...);
  if (!ui.Codec.checkFormatSupport(VK_FORMAT_BC1_RGB_UNORM_BLOCK)) { //Probably a different name
    return;
  }
  var ui.image = await ui.decodeCompressedTexture(data, width: 100, height: 100, format: VK_FORMAT_BC1_RGB_UNORM_BLOCK);
 
```

I'm not sure what iOS support looks like. We'd want to avoid having to write a support library ourselves or adding more dependencies.

**Comment by jonahwilliams** [comment]
iOS also has support for many of these compressed formats: https://developer.apple.com/documentation/metal/mtlpixelformat/mtlpixelformatastc_6x6_hdr?language=objc

**Comment by jonahwilliams** [comment]
As a starting point, we could just take the intersection of supported formats.

**Comment by jonahwilliams** [comment]
And the iOS support table is here: https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf

**Comment by chinmaygarde** [comment]
Exposing support for compressed texture formats as described is necessary, but I don't believe is sufficient in making a usable system. Especially given the myriad of formats that GPUs can natively support and the offline tooling necessary.

I am in favor of one of the [supercompressed formats in a standardized container like ktx2](https://github.com/BinomialLLC/basis_universal). In Dart code, the user can specify a file in a KTX2 container and we can do some minor conversion to a known supported format at runtime.

**Comment by chinmaygarde** [comment]
For reference, [this is the guidance on Android](https://developer.android.com/guide/playcore/asset-delivery/texture-compression#background). Also, Unity also does the transcoding like I described in an earlier comment. I am not sure if they use the open standard though. But from the perspective of an easy to use system, some sort of transcoding would be necessary.

**Comment by chinmaygarde** [comment]
More guidance on how other engines reason about a [usable way to work with compressed textures](https://doc.babylonjs.com/features/featuresDeepDive/materials/using/ktx2Compression). This also uses KTX2 for the container.

**Comment by jonahwilliams** [comment]
We don't need to build an end to end usable system though, that can be done as a package/community project. The only thing the community can't do is add engine support

**Comment by time4tea** [comment]
It would be very interesting to be able to load compressed textures directly from KTX file, and to know which texture format would be preferred from a list of possibles - for example S3TC/DXT1...

**Comment by bdero** [comment]
Folding in design detail from #187072, which I filed as a duplicate of this issue and am now closing. The Flutter GPU surface work that depends on this is tracked in #187073.

Suggested scope, grouped into three hardware families:

- BC (S3TC, RGTC, BPTC): BC1 to BC7. Desktop GPUs.
- ETC2 and EAC. Mobile, and any OpenGL ES 3.0 or WebGL2 context.
- ASTC LDR. Modern mobile and some desktop.

Skip these initially: PVRTC (deprecated, Apple only, being removed from WebGL), ETC1 (legacy, superseded by ETC2), ASTC HDR, and BC6H HDR. They can be added later behind the same query.

Work items:

1. Add the compressed `PixelFormat` values for the three families.
2. Make descriptor math block-aware. Add bytes-per-block and block dimensions to `formats.h`, and update `GetByteSizeOfBaseMipLevel` and `GetBytesPerRow` to round to block extents.
3. Add a compressed upload path: `glCompressedTexSubImage2D` (GLES), `vkCmdCopyBufferToImage` with block extents (Vulkan), and `replaceRegion` with compressed formats (Metal).
4. Enforce usage restrictions: sample plus copy-destination only. No render target, no storage, no transient storage mode, and no runtime mip generation.
5. Map the formats in each backend, gated by capability detection: Vulkan feature flags (`textureCompressionBC`, `textureCompressionETC2`, `textureCompressionASTC_LDR`), GLES extension strings, and Metal GPU family checks.

Capability query API on `impeller::Capabilities`:

```cpp
enum class TextureCompressionType {
  kBC,    // Desktop. BC1 to BC7.
  kETC2,  // Mobile, GLES3, WebGL2.
  kASTC,  // Modern mobile, some desktop.
};

// True if the device supports a whole compressed family.
virtual bool SupportsTextureCompression(TextureCompressionType type) const = 0;

// Per-format, per-usage query. Covers compressed and uncompressed formats.
virtual bool SupportsTextureFormat(PixelFormat format,
                                   TextureUsage usage) const = 0;
```

`SupportsTextureFormat` is the general primitive. `SupportsTextureCompression` is a convenience for the common "can I use this family at all" check.

No single compressed format works everywhere, which is why the query is required. Coverage is near-universal per tier: ETC2 and EAC are guaranteed on OpenGL ES 3.0 and WebGL2 and present on all mobile GPUs and Apple Silicon, BC is effectively guaranteed on desktop, and ASTC covers modern mobile. Together the three families cover essentially every target, and a single asset reaches all of them through KTX2 or Basis Universal transcoding at the asset layer, with uncompressed RGBA as the final fallback.
